File: src/_cpp/TowerDefenseBase/scenes/gamegs/MenuDig.cpp
//
//  IslandDefense
//
//  Created by Vladimir Tolmachev on 27.09.14.
//
//
#include "MenuDig.h"
#include "ScoreCounter.h"
#include "GameGS.h"
#include "Tower.h"
#include "Tutorial.h"
#include "configuration.h"

NS_CC_BEGIN


MenuDig::MenuDig()
{}

MenuDig::~MenuDig()
{
	ScoreCounter::shared().observer( kScoreLevel ).remove( _ID );
}

bool MenuDig::init()
{
	ScrollMenu::init();
	NodeExt::init();

	NodeExt::load( "ini/gamescene", "menudig.xml" );

	_dig = getItemByName( "dig" );
	_digUn = getItemByName( "dig_un" );
	_confirm = getItemByName( "confirm" );
	_confirmUn = getItemByName( "confirm_un" );

	_dig->setCallback( std::bind( &MenuDig::activate, this, std::placeholders::_1, true ) );
	_digUn->setCallback( std::bind( &MenuDig::activate, this, std::placeholders::_1, false ) );
	_confirm->setCallback( std::bind( &MenuDig::confirmSelect, this, std::placeholders::_1, true ) );
	_confirmUn->setCallback( std::bind( &MenuDig::confirmSelect, this, std::placeholders::_1, false ) );

	int cost = mlTowersInfo::shared().getCostFotDig();
	_dig->getChildByName<Label*>( "cost" )->setString( intToStr( cost ) );
	_digUn->getChildByName<Label*>( "cost" )->setString( intToStr( cost ) );

	setVisible( false );
	return true;
}

void MenuDig::appearance()
{
	setVisible( true );
	setEnabled( true );
	runEvent( "appearance" );
	ScoreCounter::shared().observer( kScoreLevel ).add( _ID, std::bind( &MenuDig::onChangeMoney, this, std::placeholders::_1 ) );
	_confirm->setVisible( false );
	_confirmUn->setVisible( false );
	_dig->setVisible( false );
	_digUn->setVisible( false );
	onChangeMoney( ScoreCounter::shared().getMoney( kScoreLevel ) );
	scheduleUpdate();
}

void MenuDig::disappearance()
{
	setEnabled( false );
	ScoreCounter::shared().observer( kScoreLevel ).remove( _ID );
	runEvent( "disappearance" );
	_confirm->setVisible( false );
	_confirmUn->setVisible( false );
	_dig->setVisible( false );
	_digUn->setVisible( false );
	unscheduleUpdate();
}

void MenuDig::setClickPoint( const Point & point )
{
	_clickedPoint = point;
	update( 0 );
	//setPosition( point );
}

void MenuDig::activate( Ref * sender, bool avalabledButton )
{
	if( isEnabled() == false )return;
	if( avalabledButton )
		_confirm->setVisible( true );
	else
		_confirmUn->setVisible( true );
	//onChangeMoney( ScoreCounter::shared().getMoney( kScoreLevel ) );
	runEvent( "onclick" );
}

void MenuDig::confirmSelect( Ref * sender, bool avalabledButton )
{
	if( isEnabled() == false )return;
	disappearance();
	if( avalabledButton )
	{
		auto place = GameGS::getInstance()->getTowerPlaceInLocation( _clickedPoint );
		GameGS::getInstance()->getGameBoard().activateTowerPlace( place );
		GameGS::getInstance()->resetSelectedPlace();
	}
	else
	{
        if (k::configuration::useInapps) {
            TutorialManager::shared( ).dispatch( "level_haventgear_dig" );
        }
	}
	runEvent( "onconfirm" );
}

void MenuDig::onChangeMoney( int money )
{
	int cost = mlTowersInfo::shared().getCostFotDig();

	bool availabled = cost <= money;
	if( _confirm->isVisible() || _confirmUn->isVisible() )
	{
		_confirm->setVisible( availabled );
		_confirmUn->setVisible( !availabled );
	}
	_dig->setVisible( availabled );
	_digUn->setVisible( !availabled );
}

void MenuDig::update( float )
{
	auto point = GameGS::getInstance()->getMainLayer()->convertToWorldSpace( _clickedPoint );
	setPosition( point );
}





NS_CC_END

File: src/_eu/scenes/gamegs/MenuDig.js
/******************************************************************************
 * Copyright 2014-2016 Vladimir Tolmachev
 * Copyright 2016 Visionarity AG
 * Vladimir Tolmachev and Visionarity AG have unlimited commercial
 * licenses for commercial use and customization
 *
 * Author: Vladimir Tolmachev (tolm_vl@hotmail.com)
 * Ported C++ to Javascript: Visionarity AG / Vladimir Tolmachev
 * Project: Island Defense (JS)
 * If you received the code not from the author, please contact us
 ******************************************************************************/

//Define namespace
var EU = EU || {};


EU.MenuDig = EU.ScrollMenu.extend( // public NodeExt
    {
        /** @type {MenuItem} */ _dig: null,
        /** @type {MenuItem} */ _digUn: null,
        /** @type {MenuItem} */ _confirm: null,
        /** @type {MenuItem} */ _confirmUn: null,
        /** @type {Point} */ _clickedPoint: null,
        /** @type {boolean} */ _waitConfirm: null,


        onExit: function () {
            EU.ScoreCounter.observer(EU.kScoreLevel).remove(this.__instanceId);
            this._super();
        },

        ctor: function () {
            this._super();
            this._waitConfirm = false;

            this.initExt();

            this.load_str_n_str("ini/gamescene", "menudig.xml");

            this._dig = this.getItemByName("dig");
            this._digUn = this.getItemByName("dig_un");
            this._confirm = this.getItemByName("confirm");
            this._confirmUn = this.getItemByName("confirm_un");

            this._dig.setCallback(this.activateDig.bind(this, true));
            this._digUn.setCallback(this.activateDig.bind(this, false));
            this._confirm.setCallback(this.confirmSelect.bind(this, true));
            this._confirmUn.setCallback(this.confirmSelect.bind(this, false));

            var cost = EU.mlTowersInfo._digcost;
            this._dig.getChildByName("cost").setString(( cost ));
            this._digUn.getChildByName("cost").setString(( cost ));

            this.setVisible(false);
            return true;
        },

        appearance: function () {
            this._waitConfirm = false;
            this.setVisible(true);
            this.setEnabled(true);
            this.runEvent("appearance");
            EU.ScoreCounter.observer(EU.kScoreLevel).add(this.__instanceId, this.onChangeMoney, this);
            this._confirm.setVisible(false);
            this._confirmUn.setVisible(false);
            this._dig.setVisible(false);
            this._digUn.setVisible(false);
            this.onChangeMoney(EU.ScoreCounter.getMoney(EU.kScoreLevel));
            this.scheduleUpdate();
        },

        disappearance: function () {
            this.setEnabled(false);
            EU.ScoreCounter.observer(EU.kScoreLevel).remove(this.__instanceId);
            this.runEvent("disappearance");
            this._confirm.setVisible(false);
            this._confirmUn.setVisible(false);
            this._dig.setVisible(false);
            this._digUn.setVisible(false);
            this.unscheduleUpdate();
        },

        setClickPoint: function (/**@type {cc.p} */ point) {
            this._clickedPoint = point;
            this.update(0);
            //setPosition( point );
        },

        activateDig: function (/**@type {bool} */ availableButton) {
            if (this.isEnabled() == false)return;
            if (availableButton) {
                this._confirm.setVisible(true);
                this._dig.setVisible(false);
            }
            else {
                this._confirmUn.setVisible(true);
                this._digUn.setVisible(false);
            }
            //onChangeMoney( EU.ScoreCounter.getMoney( EU.kScoreLevel ) );
            this.runEvent("onclick");
            this._waitConfirm = true;
        },

        confirmSelect: function (/**@type {bool} */ availableButton) {
            if (this.isEnabled() == false)return;
            this.disappearance();
            if (availableButton) {
                var place = EU.GameGSInstance.getTowerPlaceInLocation(this._clickedPoint);
                EU.GameGSInstance.getGameBoard().activateTowerPlace(place);
                EU.GameGSInstance.resetSelectedPlace();
            }
            else {
                if (EU.k.useInapps) {
                    EU.TutorialManager.dispatch("level_haventgear_dig");
                }
            }
            this.runEvent("onconfirm");
            this._waitConfirm = false;
        },

        onChangeMoney: function (/**@type {Number} */ money) {
            var cost = EU.mlTowersInfo._digcost;

            var availabled = cost <= money;
            if (this._confirm.isVisible() || this._confirmUn.isVisible()) {
                this._confirm.setVisible(availabled);
                this._confirmUn.setVisible(!availabled);
            }
            if (this._waitConfirm == false) {
                this._dig.setVisible(availabled);
                this._digUn.setVisible(!availabled);
            }
        },

        update: function () {
            var point = EU.GameGSInstance.getMainLayer().convertToWorldSpace(this._clickedPoint);
            this.setPosition(point);
        }


    });

EU.NodeExt.call(EU.MenuDig.prototype);
